src/main/java/org/opentripplanner/routing/edgetype/StreetEdge.java

```diff
@@ -403,31 +411,93 @@ else if (
-                            forkState.addToExistingResultChain(editorCar.makeState());
-                            return forkState; // return both in-car and out-of-car states
-                        } else {
-                            // if the non-car state is non traversable or something, return just the car state
-                            return editorCar.makeState();
-                        }
+                        state.addToExistingResultChain(editorCar.makeState());
+                        return state; // return both in-car and out-of-car states
                     } else {
                         // if the non-car state is non traversable or something, return just the car state
                         return editorCar.makeState();
                     }
                 }
             }
+        } else if (options.allowVehicleRental) {
+            // Irrevocable transition from using rented vehicle to walking.
+            // Final Vehicle check needed to prevent infinite recursion.
+            if (
+                s0.isVehicleRenting() &&
+                    !getPermission().allows(TraverseMode.MICROMOBILITY) &&
+                    currMode == TraverseMode.MICROMOBILITY &&
+                    // if in "arrive by" mode, the search is progressing backwards while in a vehicle rental state
+                    // before encountering a vehicle rental pickup station.  Therefore, this search cannot proceed.
+                    !options.arriveBy &&
+                    s0.isVehicleRentalDropoffAllowed(this, false)
+            ) {
+                StateEditor editorEndedVehicleRental = doTraverse(s0, options, TraverseMode.WALK);
+                if (editorEndedVehicleRental != null) {
+                    editorEndedVehicleRental.endVehicleRenting(); // done with vehicle rental use for now
+                    editorEndedVehicleRental.incrementWeight(options.vehicleRentalDropoffCost);
+                    editorEndedVehicleRental.incrementTimeInSeconds(options.vehicleRentalDropoffTime);
+                    State endedVehicleRentalState = editorEndedVehicleRental.makeState();
+                    if (state != null) {
+                        if (endedVehicleRentalState != null) {
+                            // make the forkState be of the non-vehicle-rental mode so it's possible to build walk steps
+                            endedVehicleRentalState.addToExistingResultChain(state);
+                            return endedVehicleRentalState; // return both rented-vehicle and no-rented-vehicle states
+                        }
+                    } else {
+                        // if the rented-vehicle state is non traversable or something, return just the
+                        // ended-rented-vehicle state
+                        return endedVehicleRentalState;
+                    }
+                }
+            }
+            // possible transition to dropping off a floating vehicle when in "arrive by" mode
```

Comment:
Yeah, the utlimate idea I have is to combine the car, vehicle and bike rental stuff over here. If you notice, there currently isn't any floating bike transition code on a StreetEdge because the bike rental code always assumes that a bike will be dropped off at a rental station and not on some random street.